dask/dataframe/tests/test_dataframe.py

```diff
@@ -5109,3 +5109,38 @@ def reduce_fn(series):
         .compute()[0]
     )
     assert result == {"x": 14, "y": 64}
+
+
+@pytest.fixture
+def demo_ts():
+    return dd.demo.make_timeseries()
+
+
+def test_set_index_with_dask_dt_index_gh8359(demo_ts):
+    ## was failing, now works
+    # https://github.com/dask/dask/issues/8359
+    new_index = demo_ts.index.dt.floor("D")
+    demo_ts.set_index(new_index)
+
+
+def test_set_index_with_dask_range_index_gh8359(demo_ts):
+    ## still fails
+    # https://github.com/dask/dask/issues/8359
+    df2 = demo_ts.reset_index()  # should be an IntIndex
+    new_index = df2.index
+    demo_ts.set_index(new_index)
```

Comment:
It's probably helpful to add an `assert_eq` to check that the resulting dataframe is equivalent to the pandas one.